source/iot/DeviceDefender.cpp

```diff
@@ -0,0 +1,210 @@
+                      config.taskConfig.event_loop,
+                      config.taskConfig.report_format,
+                      config.taskConfig.task_period_ns,
+                      config.taskConfig.netconn_sample_period_ns,
+                      DeviceDefenderV1ReportTask::s_onDefenderV1TaskCancelled,
+                      this},
+                  m_lastError(0)
+            {
+            }
+
+            DeviceDefenderV1ReportTask::DeviceDefenderV1ReportTask(DeviceDefenderV1ReportTask &&toMove) noexcept
+                : OnDefenderV1TaskCancelled(std::move(toMove.OnDefenderV1TaskCancelled)),
+                  cancellationUserdata(toMove.cancellationUserdata), m_allocator(toMove.m_allocator),
+                  m_status(toMove.m_status), m_taskConfig(std::move(toMove.m_taskConfig)),
+                  m_owningTask(toMove.m_owningTask), m_lastError(toMove.m_lastError)
+            {
+                toMove.OnDefenderV1TaskCancelled = nullptr;
+                toMove.cancellationUserdata = nullptr;
+                toMove.m_allocator = nullptr;
+                toMove.m_status = DeviceDefenderV1ReportTaskStatus::Stopped;
+                toMove.m_taskConfig = {0};
+                toMove.m_owningTask = nullptr;
+                toMove.m_lastError = AWS_ERROR_UNKNOWN;
+            }
+
+            DeviceDefenderV1ReportTask &DeviceDefenderV1ReportTask::operator=(
+                DeviceDefenderV1ReportTask &&toMove) noexcept
+            {
+                OnDefenderV1TaskCancelled = std::move(toMove.OnDefenderV1TaskCancelled);
+                cancellationUserdata = toMove.cancellationUserdata;
+                m_allocator = toMove.m_allocator;
+                m_status = toMove.m_status;
+                m_taskConfig = std::move(toMove.m_taskConfig);
+                m_owningTask = toMove.m_owningTask;
+                m_lastError = toMove.m_lastError;
+
+                toMove.OnDefenderV1TaskCancelled = nullptr;
+                toMove.cancellationUserdata = nullptr;
+                toMove.m_allocator = nullptr;
+                toMove.m_status = DeviceDefenderV1ReportTaskStatus::Stopped;
+                toMove.m_taskConfig = {0};
+                toMove.m_owningTask = nullptr;
+                toMove.m_lastError = AWS_ERROR_UNKNOWN;
+
+                return *this;
+            }
+
+            DeviceDefenderV1ReportTaskStatus DeviceDefenderV1ReportTask::GetStatus() noexcept { return this->m_status; }
+
+            void DeviceDefenderV1ReportTask::StartTask() noexcept
+            {
+                if (this->GetStatus() == DeviceDefenderV1ReportTaskStatus::Ready)
+                {
+
+                    this->m_owningTask = aws_iotdevice_defender_v1_report_task(this->m_allocator, &this->m_taskConfig);
+
+                    if (this->m_owningTask == nullptr)
+                    {
+                        this->m_lastError = aws_last_error();
+                        this->m_status = DeviceDefenderV1ReportTaskStatus::Failed;
```

Comment:
Question for reviewers: Is setting a failed status for the task and triggering the callback preferable to throwing en exception?

Reply:
Ok, that makes sense. It looks like we expect users to [check for return codes](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/master/samples/mqtt/basic_pub_sub/main.cpp#L471) on methods to see if the method returned with an error code. I'll remove the Failed state and update this to raise an error.